linker/shared/src/main/scala/org/scalajs/linker/backend/javascript/JSBuilders.scala

```diff
@@ -9,46 +9,35 @@
 
 package org.scalajs.linker.backend.javascript
 
-import scala.annotation.tailrec
-
-import scala.collection.mutable
+import org.scalajs.ir.Position
 
 import java.io._
-import java.util.regex.Pattern
-import java.net.{ URI, URISyntaxException }
-
-import org.scalajs.ir.Position
-import org.scalajs.io._
+import java.net.URI
 
 /** An abstract builder taking IR or JSTrees */
-trait JSTreeBuilder {
+trait JSBuilder {
   /** Add a JavaScript tree representing a statement.
    *  The tree must be a valid JavaScript tree (typically obtained by
    *  desugaring a full-fledged IR tree).
    */
   def addJSTree(tree: Trees.Tree): Unit
 
+  /** Add a chunk of JavaScript code. */
+  def addTopLevelStatement(originalLocation: URI, code: String): Unit
+
   /** Completes the builder. */
   def complete(): Unit = ()
 }
 
 class JSFileBuilder(val name: String,
-    protected val outputWriter: Writer) extends JSTreeBuilder {
+    protected val outputWriter: Writer) extends JSBuilder {
   def addLine(line: String): Unit = {
     outputWriter.write(line)
     outputWriter.write('\n')
   }
 
-  def addLines(lines: Seq[String]): Unit =
-    lines.foreach(addLine)
-
-  def addFile(file: VirtualJSFile): Unit =
-    addPartsOfFile(file)(!_.startsWith("//# sourceMappingURL="))
-
-  def addPartsOfFile(file: VirtualJSFile)(selector: String => Boolean): Unit = {
-    for (line <- file.readLines() if selector(line))
-      addLine(line)
-  }
+  def addTopLevelStatement(originalLocation: URI, code: String): Unit =
+    outputWriter.write(code)
```

Comment:
If `code` does not end with a `\n`, this method should add an addition `\n`. Otherwise, the override in `JSFileBuilderWithSourceMapWriter` will not produce a valid source map, since with its `.split('\n')` and `nextLine()` for each line, it only works correctly if the `code` ends with a `\n`.

Also care must be taken to handle the `code.isEmpty` case consistently as well.

IMO adding the `\n` is better than fixing the code in `JSFileBuilderWithSourceMapWriter`, because a "statement" might not end with a trailing `;` and rely on ASI to be correct.

It wouldn't hurt to add comments in both methods that stress that they must be in sync about this.